java-frontend/src/test/java/org/sonar/java/SonarComponentsTest.java

```diff
@@ -409,6 +418,93 @@ void jsp_classpath_should_include_plugin() throws Exception {
     assertThat(jspClassPath).containsExactly(plugin.getAbsolutePath(), someJar.getAbsolutePath());
   }
 
+  @Nested
+  class Logging {
+    private final DecimalFormat formatter = new DecimalFormat("00");
+
+    private final SensorContextTester context = SensorContextTester.create(new File(""));
+    private final DefaultFileSystem fs = context.fileSystem();
+    private final Configuration settings = context.config();
+
+    private final ClasspathForMain javaClasspath = new ClasspathForMain(settings, fs);
+    private final ClasspathForTest javaTestClasspath = new ClasspathForTest(settings, fs);
+
+    private SonarComponents sonarComponents;
+
+    @org.junit.Rule
+    public LogTester logTester = new LogTester();
+
+    @BeforeEach
+    void beforeEach() {
+      sonarComponents = new SonarComponents(null, fs, javaClasspath, javaTestClasspath, null);
+      logTester.clear();
+    }
+
+    @Test
+    void log_only_50_undefined_types() {
+      String source = generateSource(26);
+
+      // artificially populated the semantic errors with 26 unknown types and 52 errors
+      sonarComponents.collectUndefinedTypes(((JavaTree.CompilationUnitTreeImpl) JParserTestUtils.parse(source)).sema.undefinedTypes());
+
+      // triggers log
+      sonarComponents.logUndefinedTypes();
+
+      List<String> debugLogs = logTester.logs(LoggerLevel.DEBUG);
+      assertThat(debugLogs)
+        .hasSize(2)
+        .startsWith("Unresolved imports/types have been detected during analysis. Logging the 50 first:");
+      String list = debugLogs.get(1);
+
+      assertThat(list)
+        .endsWith("- ...")
+        .doesNotContain("- Y cannot be resolved to a type")
+        .doesNotContain("- Z cannot be resolved to a type");
+      for (int i = 0; i < 26; i++) {
```

Comment:
It's just a matter of taste, so what do you think about this variant:

```
      IntStream.rangeClosed(0, 25).forEach(i -> {
          char typeName = (char) ('A' + i);
          assertThat(list).contains(String.format("- The import org.package%s cannot be resolved", formatter.format(i + 1)));
          if (((char) ('A' + i)) < 'Y') {
            assertThat(list).contains(String.format("- %c cannot be resolved to a type", typeName));
          }
        });
```

Reply:
I definitely feel better letting the test without requiring to streams there, so I'm going to keep as it is for now! But thanks for the suggestion!